File: JavaScript/fraction.js
class Fraction {

  constructor(...args) {
    this.set(...args);
    this.epsilon = 5e-6;
  }

  get numerator() { return this.numerator_; }
  get denominator() { return this.denominator_; }

/*  set epsilon(eps) { this.epsilon = eps; }
  get epsilon() { return this.epsilon; }*/

  static epsilon = 5e-6;
  static loops = 0;
  static gcd(a,b) {
    while(b!= 0) {
      let t = b;
      b = a % b;
      a = t;
    }
    return a;
  }

  static set_num_denom(f,n,d) {
    if(d < 0) {
      n=-n;
      d=-d;
    }
    let divisor=Fraction.gcd(Math.abs(n),d)
    f.numerator_=n/divisor;
    f.denominator_=d/divisor;
  }

  static set_real(f,d) {
    var hm2=0,hm1=1,km2=1,km1=0;
    var h,k;
    var v = d;
    Fraction.loops=0;
    while(true) {
      var a=Math.floor(v);
      h=a*hm1 + hm2;
      k=a*km1 + km2;
//      print(a,h,k);
      if(Math.abs(d - h/k) < Fraction.epsilon) {
        break;
      }
      v = 1.0/(v-a);
      hm2=hm1;
      hm1=h;
      km2=km1;
      km1=k;
      Fraction.loops++;
    }
    if(k<0) {
      k=-k;
      h=-h;
    }
    f.numerator_=h;
    f.denominator_=k;
  }

  /* # of arguments
    0 set to default values
    1 A number (int or float), string, or Fraction
    2 Two integers to set numerator and denominator
    3 Three integers to set as mixed fraction (whole, numerator, and denominator)
  */
  set(arg1,arg2,arg3)
  {
    if(arguments.length == 0) {
      this.numerator_=0; this.denominator_=1;
    } else if(arguments.length == 1) {
      if(typeof(arg1)=="object") { //has to be anoter fraction object
        this.numerator_=arg1.numerator_;
        this.denominator_=arg1.denominator_;
      } else { // a number
        Fraction.set_real(this,arg1);
      }
    } else if(arguments.length == 2) {
      Fraction.set_num_denom(this,arg1,arg2)
    } else if(arguments.length == 3) {
      Fraction.set_num_denom(this,arg1*arg3+(arg1 < 0 ? -1 : 1)*arg2,arg3);
    } else {
      /* Error */
    }
  }

  plus(b) {
    this.set(this.numerator_*b.denominator_ + this.denominator_*b.numerator_,this.denominator_*b.denominator_);
  }

  minus(b) {
    this.set(this.numerator_*b.denominator_ - this.denominator_*b.numerator_,this.denominator_*b.denominator_);
  }

  times(b) {
    this.set(this.numerator_*b.numerator_,this.denominator_*b.denominator_);
  }

  divided_by(b) {
    this.set(this.numerator_*b.denominator_,this.denominator_*b.numerator_);
  }

  static fraction_plus_fraction(a,b)
  {
    var c = new Fraction();
    c.numerator_=a.numerator_;
    c.denominator_=a.denominator_;
    c.plus(b);
    return c;
  }

  static fraction_minus_fraction(a,b)
  {
    var c = new Fraction();
    c.numerator_=a.numerator_;
    c.denominator_=a.denominator_;
    c.minus(b);
    return c;
  }

  static fraction_times_fraction(a,b)
  {
    var c = new Fraction();
    c.numerator_=a.numerator_;
    c.denominator_=a.denominator_;
    c.times(b);
    return c;
  }

  static fraction_divided_by_fraction(a,b)
  {
    var c = new Fraction();
    c.numerator_=a.numerator_;
    c.denominator_=a.denominator_;
    c.divided_by(b);
    return c;
  }

  equal(rhs)
  {
    return Fraction.cmp(this,rhs) == 0;
  }

  static fraction_equal_fraction(lhs,rhs) {
    return lhs.equal(rhs);
  }

  not_equal(rhs)
  {
    return Fraction.cmp(this,rhs) != 0;
  }

  static fraction_not_equal_fraction(lhs,rhs)
  {
    return lhs.not_equal(rhs);
  }

  less_than(rhs)
  {
    return Fraction.cmp(this,rhs) < 0;
  }

  static fraction_less_than_fraction(lhs,rhs)
  {
    return lhs.less_than(rhs);
  }

  less_than_equal(rhs)
  {
    return Fraction.cmp(this,rhs) <= 0;
  }

  static fraction_less_than_equal_fraction(lhs,rhs)
  {
    return lhs.less_than_equal(rhs);
  }

  greater_than(rhs)
  {
    return Fraction.cmp(this,rhs) > 0;
  }

  static fraction_greater_than_fraction(lhs,rhs)
  {
    return lhs.greater_than(rhs);
  }

  greater_than_equal(rhs)
  {
    return Fraction.cmp(this,rhs) >= 0;
  }

  static fraction_greater_than_equal_fraction(lhs,rhs)
  {
    return lhs.greater_than_equal(rhs);
  }

  static cmp(lhs,rhs) {
    return lhs.numerator_*rhs.denominator_ - rhs.numerator_*lhs.denominator_;
  }

  round(d)
  {
    if(d<this.denominator_) {
      this.set(Math.round(this.numerator_*d/this.denominator_),Math.round(d));
    }
    return this;
  }
  toString() {
    if(this.denominator_ == 1) {
      return this.numerator_.toString();
    }
    return this.numerator_ + "/" + this.denominator_;
  }

  toMixedString() {
    if(this.numerator_ <= this.denominator_) {
      return this.toString();
    }
    var whole=Math.floor(this.numerator_/this.denominator_);
    var n = Math.abs(this.numerator_) % this.denominator_;
    if(n==0) {
      return whole.toString();
    }
    return whole+" "+n+"/"+this.denominator_;
  }

  valueOf() {
    return this.numerator_/this.denominator_;
  }
}

module.exports = Fraction;

File: Pascal/fraction.pas
(*
		Copyright (C) 2019-2021  by Terry N Bezue

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*)

unit fraction;
interface
uses sysutils,math;

type TFraction = object
  private
    _numerator,_denominator : longint;
//    _epsilon: real; static;
    procedure reduce();
  public
    function epsilon: real; static;
    procedure epsilon(e: real); static;
    function gcd(a,b: longint): longint; static;
    function numerator: longint;
    function denominator: longint;
(*    constructor create; overload;
    constructor create(n: longint); overload;
    constructor create(n,d: longint); overload;
    constructor create(w,n,d: longint); overload;
    constructor create(r: real); overload;
    constructor create(s: string); overload; *)
    procedure fset(num: longint ); overload;
    procedure fset(num,denom: longint); overload;
    procedure fset(whole,num,denom: longint); overload;
    procedure fset(d: real); overload;
    function Round(denom: longint): TFraction;
//    function ToStr: string;
//    function ToMixedStr: string;
end;

type TMixedFraction = object(TFraction)
end;

type
  PFraction = ^TFraction;
  PMixedFraction = ^TMixedFraction;

operator := (r: real) f:TFraction;
operator := (a: array of longint) f: TFraction;

operator := (r: real) mf: TMixedFraction;
operator := (a: array of longint) f: TMixedFraction;
operator := (f: TFraction) mf: TMixedFraction;

operator = (lhs,rhs: TFraction) b : boolean;
operator <> (lhs,rhs: TFraction) b : boolean;
operator < (lhs,rhs: TFraction) b : boolean;
operator <= (lhs,rhs: TFraction) b : boolean;
operator > (lhs,rhs: TFraction) b : boolean;
operator >= (lhs,rhs: TFraction) b : boolean;

operator = (lhs: TFraction; rhs: real) b : boolean;
operator <> (lhs: TFraction; rhs: real) b : boolean;
operator < (lhs: TFraction; rhs: real) b : boolean;
operator <= (lhs: TFraction; rhs: real) b : boolean;
operator > (lhs: TFraction; rhs: real) b : boolean;
operator >= (lhs: TFraction; rhs: real) b : boolean;

operator + (lhs,rhs: TFraction) f: TFraction;
operator - (lhs,rhs: TFraction) f: TFraction;
operator * (lhs,rhs: TFraction) f: TFraction;
operator / (lhs,rhs: TFraction) f: TFraction;
operator ** (lhs,rhs: TFraction) f: TFraction;

operator + (lhs: TFraction; rhs: real) f: TFraction;
operator - (lhs: TFraction; rhs: real) f: TFraction;
operator * (lhs: TFraction; rhs: real) f: TFraction;
operator / (lhs: TFraction; rhs: real) f: TFraction;
operator ** (lhs: TFraction; rhs: real) f: TFraction;

operator + (lhs: real; rhs: TFraction) r: real;
operator - (lhs: real; rhs: TFraction) r: real;
operator * (lhs: real; rhs: TFraction) r: real;
operator / (lhs: real; rhs: TFraction) r: real;
operator ** (lhs: real; rhs: TFraction) r: real;

operator explicit (f: TFraction) s: string;
operator explicit (mf: TMixedFraction) s: string;

var
  loops: longint;
  _epsilon: real = 5e-6;

Implementation

function TFraction.gcd(a,b: longint): longint; static;
  var
    t: longint;
  begin
    while (b <> 0) do
    begin
      t := b;
      b := a mod b;
      a := t;
    end;
    gcd := a;
  end;

function TFraction.epsilon: real; static;
  begin
    epsilon := _epsilon;
  end;

procedure TFraction.epsilon(e: real); static;
  begin
    _epsilon := e;
  end;

function TFraction.numerator: longint;
  begin
    numerator := _numerator;
  end;

function TFraction.denominator: longint;
  begin
     denominator := _denominator;
  end;

procedure TFraction.reduce();
  var
    divisor: longint;
begin
  if _denominator < 0 then
  begin
    _numerator := -_numerator;
    _denominator := -_denominator;
  end;

  divisor := gcd(abs(_numerator),_denominator);
  if divisor <> 1  then
  begin
    _numerator := _numerator div divisor;
    _denominator := _denominator div divisor;
  end;
end;
(*
constructor TFraction.create();
  begin
    self._numerator := 0;
    self._denominator := 1;
  end;

constructor TFraction.create(n: longint);
  begin
    _numerator := n;
    _denominator := 1;
  end;


constructor TFraction.create(n,d: longint);
  begin
    fset(n,d);
  end;

constructor TFraction.create(w,n,d: longint);
  begin
    fset(w,n,d);
  end;

constructor TFraction.create(r: real);
  begin
    fset(r);
  end;

constructor TFraction.create(s:string);
  begin;
  end;
*)
procedure TFraction.fset(num: longint);
begin
  _numerator := num;
  _denominator := 1;
end;

procedure TFraction.fset(num,denom: longint);
begin
  _numerator := num;
  _denominator := denom;
  Reduce();
end;

procedure TFraction.fset(whole,num,denom: longint);
  var
    sign: longint;
begin
  sign := 1;
  if whole < 0 then
  begin
    sign := -sign;
    whole := -whole;
  end;
  if num < 0 then
  begin
    sign := -sign;
    num := -num;
  end;
  if denom < 0 then
  begin
    sign := -sign;
    denom := -denom;
  end;
  fset(sign*(whole*denom+num),denom);
end;

procedure TFraction.fset(d: real);
var
  a,hm2,hm1,h,km2,km1,k: longint;
  v: real;
begin
  hm2:=0; hm1:=1; km2:=1; km1:=0;
  v:=d;
  loops := 0;
  while true do
  begin
    a:=trunc(v);
    h:= a*hm1 + hm2;
    k:=a*km1 + km2;
    if (abs(d - h/k) < _epsilon) then
      break;
    loops := loops + 1;
    v := 1.0/(v - a);
    hm2:=hm1;
    hm1:=h;
    km2:=km1;
    km1:=k;
  end;
  if(k<0) then
  begin
    k := -k;
    h := -h;
  end;
  _numerator:=h;
  _denominator:=k;
end;

function TFraction.Round(denom: longint): TFraction;
begin
  Round.fset(trunc(denom*_numerator/_denominator + 0.5),denom);
end;

operator + (lhs,rhs: TFraction) f: TFraction;
  begin
    f.fset(lhs._numerator*rhs._denominator + rhs._numerator*lhs._denominator,lhs._denominator*rhs._denominator);
  end;

operator - (lhs,rhs: TFraction) f: TFraction;
  begin
    f.fset(lhs._numerator*rhs._denominator - rhs._numerator*lhs._denominator,lhs._denominator*rhs._denominator);
  end;


operator * (lhs,rhs: TFraction) f: TFraction;
  begin
    f.fset(lhs._numerator*rhs._numerator,lhs._denominator*rhs._denominator);
  end;


operator / (lhs,rhs: TFraction) f: TFraction;
  begin
    f.fset(lhs._numerator*rhs._denominator,lhs._denominator*rhs._numerator);
  end;

operator ** (lhs,rhs: TFraction) f: TFraction;
  begin
    f := (lhs._numerator/lhs._denominator)**(rhs._numerator/rhs._denominator);
  end;

operator + (lhs: TFraction; rhs: real) f: TFraction;
  var
    frhs: TFraction;
  begin
    frhs.fset(rhs);
    f := lhs + frhs;
  end;

operator - (lhs: TFraction; rhs: real) f: TFraction;
  var
    frhs: TFraction;
  begin
    frhs.fset(rhs);
    f := lhs - frhs;
  end;

operator * (lhs: TFraction; rhs: real) f: TFraction;
  var
    frhs: TFraction;
  begin
    frhs.fset(rhs);
    f := lhs * frhs;
  end;

operator / (lhs: TFraction; rhs: real) f: TFraction;
  var
    frhs: TFraction;
  begin
    frhs.fset(rhs);
    f := lhs / frhs;
  end;

operator ** (lhs: TFraction; rhs: real) f: TFraction;
  var
    frhs: TFraction;
  begin
    frhs.fset(rhs);
    f := lhs ** frhs;
  end;

operator + (lhs: real; rhs: TFraction) r: real;
  begin
    r := lhs + rhs.numerator/rhs.denominator;
  end;

operator - (lhs: real; rhs: TFraction) r: real;
  begin
    r := lhs - rhs.numerator/rhs.denominator;
  end;

operator * (lhs: real; rhs: TFraction) r: real;
  begin
    r := lhs * rhs.numerator/rhs.denominator;
  end;

operator / (lhs: real; rhs: TFraction) r: real;
  begin
    r := lhs*rhs.denominator/rhs.numerator;
  end;

operator ** (lhs: real; rhs: TFraction) r: real;
  begin
    r := lhs ** (rhs.numerator/rhs.denominator);
  end;

(*
function TFraction.ToStr: string;
begin
  if _denominator = 1 then
    ToStr := Format('%d',[_numerator])
  else
    ToStr := Format('%d/%d',[_numerator,_denominator]);
end;
*)
operator explicit(f: TFraction) s: string;
  begin
    if f.denominator = 1 then
      s := Format('%d',[f.numerator])
    else
      s := Format('%d/%d',[f.numerator,f.denominator]);
  end;
(*
function TFraction.ToMixedStr: string;
var
  whole,num: longint;
begin
  if (_denominator <> 1) or (abs(_numerator) > _denominator) then
  begin
    whole := trunc(_numerator/_denominator);
    num := abs(_numerator) - abs(whole)*_denominator;
    ToMixedStr := format('%d %d/%d',[whole,num,_denominator]);
  end
  else
    ToMixedStr := ToStr();
end;
*)
function cmp(lhs,rhs: TFraction): integer;
begin
  cmp := lhs.numerator*rhs.denominator - rhs.numerator*lhs.denominator;
end;

operator = (lhs,rhs : TFraction) b: boolean;
  begin
    b := cmp(lhs,rhs) = 0;
  end;

operator <> (lhs,rhs : TFraction) b: boolean;
  begin
    b := cmp(lhs,rhs) <> 0;
  end;

operator < (lhs,rhs : TFraction) b: boolean;
  begin
    b := cmp(lhs,rhs) < 0;
  end;

operator <= (lhs,rhs : TFraction) b: boolean;
  begin
    b := cmp(lhs,rhs) <= 0;
  end;

operator > (lhs,rhs : TFraction) b: boolean;
  begin
    b := cmp(lhs,rhs) > 0;
  end;

operator >= (lhs,rhs : TFraction) b: boolean;
  begin
    b := cmp(lhs,rhs) >= 0;
  end;

function cmp(lhs: TFraction; rhs:real): integer;
  var
    frhs: TFraction;
  begin
    frhs := rhs;
    cmp := cmp(lhs,frhs);
  end;

operator = (lhs: TFraction; rhs: real) b: boolean;
  begin
    b := cmp(lhs,rhs) = 0;
  end;

operator <> (lhs: TFraction; rhs: real) b: boolean;
  begin
    b := cmp(lhs,rhs) <> 0;
  end;

operator < (lhs: TFraction; rhs: real) b: boolean;
  begin
    b := cmp(lhs,rhs) < 0;
  end;

operator <= (lhs: TFraction; rhs: real) b: boolean;
  begin
    b := cmp(lhs,rhs) <= 0;
  end;

operator > (lhs: TFraction; rhs: real) b: boolean;
  begin
    b := cmp(lhs,rhs) > 0;
  end;

operator >= (lhs: TFraction; rhs: real) b: boolean;
  begin
    b := cmp(lhs,rhs) >= 0;
  end;

function cmp(lhs: real; rhs:TFraction): integer;
  var
    rrhs: real;
begin
  rrhs := rhs.numerator/rhs.denominator;
  if abs(lhs - rrhs) < _epsilon then
    cmp := 0
  else
    if lhs < rrhs then
      cmp := -1
    else
      cmp := 1;
end;

operator = (lhs: real; rhs: TFraction) b: boolean;
  begin
    b := cmp(lhs,rhs) = 0;
  end;

operator <> (lhs: real; rhs: TFraction) b: boolean;
  begin
    b := cmp(lhs,rhs) <> 0;
  end;

operator < (lhs: real; rhs: TFraction) b: boolean;
  begin
    b := cmp(lhs,rhs) < 0;
  end;

operator <= (lhs: real; rhs: TFraction) b: boolean;
  begin
    b := cmp(lhs,rhs) <= 0;
  end;

operator > (lhs: real; rhs: TFraction) b: boolean;
  begin
    b := cmp(lhs,rhs) > 0;
  end;

operator >= (lhs: real; rhs: TFraction) b: boolean;
  begin
    b := cmp(lhs,rhs) >= 0;
  end;

operator := (r: real) f:TFraction;
  begin
    f.fset(r);
  end;

operator := (a: array of longint) f: TFraction;
  var
    alen: longint;
    ilow: longint;
  begin
    alen := length(a);
    ilow := low(a);
    case alen of
      0: f.fset(0,1);
      1: f.fset(a[ilow]);
      2: f.fset(a[ilow],a[ilow+1]);
    else
      f.fset(a[ilow],a[ilow+1],a[ilow+2]);
    end;
  end;

operator := (r: real) mf:TMixedFraction;
  begin
    mf.fset(r);
  end;

operator := (a: array of longint) f: TMixedFraction;
  var
    alen: longint;
    ilow: longint;
  begin
    alen := length(a);
    ilow := low(a);
    case alen of
      0: f.fset(0,1);
      1: f.fset(a[ilow]);
      2: f.fset(a[ilow],a[ilow+1]);
    else
      f.fset(a[ilow],a[ilow+1],a[ilow+2]);
    end;
  end;

operator := (f: TFraction) mf: TMixedFraction;
  begin
    mf.fset(f.numerator,f.denominator);
  end;

operator explicit(mf: TMixedFraction) s: string;
  var
    whole,num: longint;
  begin
    if mf.denominator = 1 then
      s := Format('%d',[mf.numerator])
    else
      if abs(mf.numerator) < mf.denominator then
        s:= Format('%d/%d',[mf.numerator,mf.denominator])
      else
        begin
          whole := trunc(mf.numerator/mf.denominator);
          num := abs(mf.numerator) - abs(whole)*mf.denominator;
          s := format('%d %d/%d',[whole,num,mf.denominator]);
        end;
    end;
end.
